WooCommerce/Classes/ViewRelated/Products/Media/WordPressMediaLibraryPickerDataSource.swift

```diff
@@ -0,0 +1,189 @@
+import WPMediaPicker
+import Yosemite
+import WordPressShared
+
+/// Implements `WPMediaGroup` protocol that represents a collection of media items from WP Media Library to be displayed.
+///
+final class WordPressMediaLibraryMediaGroup: NSObject, WPMediaGroup {
+    private let mediaItems: [Media]
+
+    init(mediaItems: [Media]) {
+        self.mediaItems = mediaItems
+        super.init()
+    }
+
+    func name() -> String {
+        return NSLocalizedString("WordPress Media Library", comment: "Navigation bar title for WordPress Media Library image picker")
+    }
+
+    func image(with size: CGSize, completionHandler: @escaping WPMediaImageBlock) -> WPMediaRequestID {
+        return 0
+    }
+
+    func cancelImageRequest(_ requestID: WPMediaRequestID) {}
+
+    func baseGroup() -> Any {
+        return ""
+    }
+
+    func identifier() -> String {
+        return "group id"
+    }
+
+    func numberOfAssets(of mediaType: WPMediaType, completionHandler: WPMediaCountBlock? = nil) -> Int {
+        return mediaItems.count
+    }
+}
+
+/// Implements `WPMediaCollectionDataSource` that provides the media data from WP Media Library for the picker UI.
+///
+final class WordPressMediaLibraryPickerDataSource: NSObject {
+    private let siteID: Int64
+    private let syncingCoordinator: SyncingCoordinator
+
+    /// Called when the media data have changed from outside of `loadData` calls. For example, this is called when the media data for the next page return.
+    private var onDataChange: WPMediaChangesBlock?
+
+    private var mediaItems: [Media]
+
+    private lazy var mediaGroup: WPMediaGroup = WordPressMediaLibraryMediaGroup(mediaItems: mediaItems)
+
+    init(siteID: Int64) {
+        self.siteID = siteID
+        mediaItems = []
+        syncingCoordinator = SyncingCoordinator(pageSize: Constants.numberOfItemsPerPage)
+        super.init()
+
+        syncingCoordinator.delegate = self
+    }
+}
+
+extension WordPressMediaLibraryPickerDataSource: WPMediaCollectionDataSource {
+    func numberOfGroups() -> Int {
+        return 1
+    }
+
+    func group(at index: Int) -> WPMediaGroup {
+        return mediaGroup
+    }
+
+    func selectedGroup() -> WPMediaGroup? {
+        return mediaGroup
+    }
+
+    func numberOfAssets() -> Int {
+        return mediaItems.count
+    }
+
+    func media(at index: Int) -> WPMediaAsset {
+        let media = mediaItems[index]
+
+        // Since there is no paging API from the media picker library, this is where we detect whether the last item has been reached via sync coordinator.
+        syncingCoordinator.ensureNextPageIsSynchronized(lastVisibleIndex: index)
+
+        return media
+    }
+
+    func media(withIdentifier identifier: String) -> WPMediaAsset? {
+        return mediaItems.first(where: { "\($0.mediaID)" == identifier })
```

Comment:
⚠️ I'm worried that there are 2 sources of truths about what the "identifier" is. In `Media+WPMediaAsset`, we _create_ the identifier by converting the `mediaID` to `String`:

https://github.com/woocommerce/woocommerce-ios/blob/e867e0e208afa42a6ce3d3ece7cee16f8cb4ce4b/WooCommerce/Classes/ViewRelated/Products/Media/Media%2BWPMediaAsset.swift#L46-L48

And we do the same here. If one of them changes, we'll have to change both. 

I think we can use the `identifier()` method here instead: 

```suggestion
        mediaItems.first(where: { $0.identifier() == identifer })
```

Reply:
thanks for catching this! updated here: https://github.com/woocommerce/woocommerce-ios/pull/2060/commits/78a650006673d9295667089e9a7d0057ba7f4a4a